core/sdk-core/src/main/java/software/amazon/awssdk/core/progress/listener/ProgressListener.java

```diff
@@ -0,0 +1,395 @@
+ * Request execution successful!
+ * </pre>
+ */
+@SdkPublicApi
+public interface ProgressListener {
+
+    /**
+     * This method is called right after a request object is marshalled and ready to be sent to the service
+     * <p>
+     * Available context attributes:
+     * <ol>
+     *     <li>{@link Context.RequestPrepared#request()}</li>
+     *     <li>{@link Context.RequestPrepared#progressSnapshot()}</li>
+     * </ol>
+     */
+    default void requestPrepared(Context.RequestPrepared context) {
+    }
+
+    /**
+     * This method is called after the request transaction is initiated, i.e. request header is sent to the service
+     * <p>
+     * Available context attributes:
+     * <ol>
+     *     <li>{@link Context.RequestHeaderSent#request()}</li>
+     *     <li>{@link Context.RequestHeaderSent#progressSnapshot()}</li>
+     * </ol>
+     */
+    default void requestHeaderSent(Context.RequestHeaderSent context) {
+    }
+
+    /**
+     * This method is called with any additional payload bytes sent; it may be called many times per request, depending on the
+     * payload size.
+     * <p>
+     * Available context attributes:
+     * <ol>
+     *     <li>{@link Context.RequestBytesSent#request()}</li>
+     *     <li>{@link Context.RequestBytesSent#progressSnapshot()}</li>
+     * </ol>
+     */
+    default void requestBytesSent(Context.RequestBytesSent context) {
+    }
+
+    /**
+     * The service returns the response headers
+     * <p>
+     * Available context attributes:
+     * <ol>
+     *     <li>{@link Context.ResponseHeaderReceived#request()}</li>
+     *     <li>{@link Context.ResponseHeaderReceived#progressSnapshot()}</li>
+     * </ol>
+     */
+    default void responseHeaderReceived(Context.ResponseHeaderReceived context) {
+    }
+
+    /**
+     * Additional bytes received
+     * <p>
+     * Available context attributes:
+     * <ol>
+     *     <li>{@link Context.ResponseBytesReceived#request()}</li>
+     *     <li>{@link Context.ResponseBytesReceived#progressSnapshot()}</li>
+     * </ol>
+     */
+    default void responseBytesReceived(Context.ResponseBytesReceived context) {
+    }
+
+    /**
+    * Additional bytes received
+    * <p>
+    * Available context attributes:
+    * <ol>
+    *     <li>{@link Context.ResponseBytesReceived#request()}</li>
+    *     <li>{@link Context.ResponseBytesReceived#progressSnapshot()}</li>
+    * </ol>
+    */
+    default void attemptFailureResponseBytesReceived(Context.AttemptFailureResponseBytesReceived context) {
+    }
+
+    /**
+     * Successful request execution
+     * <p>
+     * Available context attributes:
+     * <ol>
+     *     <li>{@link Context.ExecutionSuccess#request()}</li>
+     *     <li>{@link Context.ExecutionSuccess#progressSnapshot()}</li>
+     *     <li>{@link Context.ExecutionSuccess#executionSuccess()} ()}</li>
+     * </ol>
+     */
+    default void executionSuccess(Context.ExecutionSuccess context) {
+    }
```

Comment:
Did we intend to include an executionFailure method as well, so that the user knows when the retries have all failed? (we'd call executionFailure after attemptFailure if we are done retrying). I see it referenced in a few places below.

Reply:
Yes, I actually meant to put it in but missed that one. Adding it in now